### PR TITLE
Implement `handleRemoveHistoryItem` for variant analysis history items

### PR DESCRIPTION
**Pull request by elenatanasoiu**
🚨 Please review this PR commit-by-commit.

We had previously added a no-op placeholder for when we attempt
to remove a variant analysis from our query history.

This adds the implementation by:
- removing the item from the query history
- cleaning up any existing result files attached to the variant analysis

NB: For remote queries, we would store all their results in a single folder.
For variant analyses, we store results per repo. The folder names are 
generated using a `cacheKey` which is built from the variant analysis ID and
the repo name. The resulting `cacheKeys` are stored in `cachedResults`.

Since we need to know the repo name when we want to delete a results 
folder,  we've had to pass in the full variant analysis object to the manager 
and call `cacheResults.delete()` for each of its scanned repos.

Also we're adding tests, because we need more of them. ❤️ 

## Checklist

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

File: extensions/ql-vscode/src/query-history.ts [hunk 1]
```diff
@@ -635,7 +636,7 @@ export class QueryHistoryManager extends DisposableObject {
     const variantAnalysisRemovedSubscription = this.variantAnalysisManager.onVariantAnalysisRemoved(async (variantAnalysis) => {
       const items = this.treeDataProvider.allHistory.filter(i => i.t === 'variant-analysis' && i.variantAnalysis.id === variantAnalysis.id);
```
Comment (norascheuch) [inline extensions/ql-vscode/src/query-history.ts:637]: Just wondering about it: can there be multiple items? With the decision to remove the historyId there should always only be one history item per variant analysis?

File: extensions/ql-vscode/src/query-history.ts [hunk 2]
```diff
@@ -812,6 +813,18 @@ export class QueryHistoryManager extends DisposableObject {
     await this.remoteQueriesManager.removeRemoteQuery(item.queryId);
   }
 
+  private async removeVariantAnalysis(item: VariantAnalysisHistoryItem): Promise<void> {
+    // Remote queries can be removed locally, but not remotely.
```
Comment (norascheuch) [inline extensions/ql-vscode/src/query-history.ts:817]:
```suggestion
    // We can remove a variant analysis locally, but not remotely.
```

File: extensions/ql-vscode/src/remote-queries/variant-analysis-results-manager.ts [hunk 3]
```diff
@@ -179,6 +179,19 @@ export class VariantAnalysisResultsManager extends DisposableObject {
     return `https://github.com/${fullName}/blob/${sha}`;
   }
 
+  public removeAnalysesResults(variantAnalysis: VariantAnalysis) {
```
Comment (norascheuch) [inline extensions/ql-vscode/src/remote-queries/variant-analysis-results-manager.ts:182]:
```suggestion
  public removeAnalysisResults(variantAnalysis: VariantAnalysis) {
```

File: extensions/ql-vscode/src/remote-queries/variant-analysis-manager.ts [hunk 4]
```diff
@@ -72,6 +72,17 @@ export class VariantAnalysisManager extends DisposableObject implements VariantA
     }
   }
 
+  public async removeVariantAnalysis(variantAnalysis: VariantAnalysis) {
+    this.variantAnalysisResultsManager.removeAnalysesResults(variantAnalysis);
```
Comment (norascheuch) [inline extensions/ql-vscode/src/remote-queries/variant-analysis-manager.ts:76]: I think this should be named `removeAnalysisResults` as it's removing multiple results for only one analysis?

Comment (norascheuch) [reply extensions/ql-vscode/src/remote-queries/variant-analysis-manager.ts:76]: I started suggesting the change but it has a couple of occurrences so it's easier to do it in the editor

File: extensions/ql-vscode/src/remote-queries/variant-analysis-manager.ts [hunk 5]
```diff
@@ -103,6 +114,14 @@ export class VariantAnalysisManager extends DisposableObject implements VariantA
...
   }
 
+  public getVariantAnalyses(): Map<number, VariantAnalysis> {
+    return this.variantAnalyses;
+  }
+
+  public async setVariantAnalysis(variantAnalysis: VariantAnalysis): Promise<void> {
+    this.variantAnalyses.set(variantAnalysis.id, variantAnalysis);
+  }
+
```
Comment (norascheuch) [inline extensions/ql-vscode/src/remote-queries/variant-analysis-manager.ts:124]: These both are only created for tests, right? .. 😬

**Review by norascheuch — COMMENTED** [review]
Left a couple of questions and comments. I don't feel good with the test changes and methods created just for tests but I suppose it's not possible to write good tests otherwise?..

File: extensions/ql-vscode/src/remote-queries/variant-analysis-manager.ts [hunk 6]
```diff
@@ -103,6 +114,14 @@ export class VariantAnalysisManager extends DisposableObject implements VariantA
     return this.variantAnalyses.get(variantAnalysisId);
   }
 
+  public getVariantAnalyses(): Map<number, VariantAnalysis> {
```
Comment (charisk) [inline extensions/ql-vscode/src/remote-queries/variant-analysis-manager.ts:117]: This is quite risky/leaky. It gives any part of the codebase access to get and edit variant analyses. If we _have to_ expose them, we should try and protect against random changes to the data by returning a copy of the real data. Perhaps using [immutablejs](https://immutable-js.com/) which is already used in the codebase.

Comment (elenatanasoiu) [reply extensions/ql-vscode/src/remote-queries/variant-analysis-manager.ts]: I've removed this public method and replaced it with a smaller one which just returns the size: https://github.com/github/vscode-codeql/pull/1655/commits/51b547b411d8afd09eaa1f3f87d18c43b61970d7

**Comment by elenatanasoiu** [comment]
Closing in favour of https://github.com/github/vscode-codeql/pull/1656

I'm going to split this into smaller PRs so it doesn't block us.

**Pull request closed without merging**